examples/d3d12/src/main.rs

```diff
@@ -0,0 +1,855 @@
+                    pResource: Some(resource.clone()),
+                    StateBefore: state_before,
+                    StateAfter: state_after,
+                    Subresource: D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES,
+                }
+                .abi(),
+            },
+        }
+    }
+
+    fn create_device(command_line: &SampleCommandLine) -> Result<(IDXGIFactory4, ID3D12Device)> {
+        if cfg!(debug_assertions) {
+            unsafe {
+                if let Ok(debug) = D3D12GetDebugInterface::<ID3D12Debug>() {
+                    debug.EnableDebugLayer();
+                }
+            }
+        }
+
+        let dxgi_factory_flags = if cfg!(debug_assertions) {
+            DXGI_CREATE_FACTORY_DEBUG
+        } else {
+            0
+        };
+
+        let dxgi_factory: IDXGIFactory4 = unsafe { CreateDXGIFactory2(dxgi_factory_flags)? };
+
+        let adapter = if command_line.use_warp_device {
+            unsafe { dxgi_factory.EnumWarpAdapter() }
+        } else {
+            get_hardware_adapter(&dxgi_factory)
+        }?;
+
+        let device = unsafe { D3D12CreateDevice(adapter, D3D_FEATURE_LEVEL_11_0)? };
+        Ok((dxgi_factory, device))
+    }
+
+    fn create_root_signature(device: &ID3D12Device) -> Result<ID3D12RootSignature> {
+        let desc = D3D12_ROOT_SIGNATURE_DESC {
+            Flags: D3D12_ROOT_SIGNATURE_FLAG_ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT,
+            ..Default::default()
+        };
+
+        unsafe {
+            let mut error = None;
+            let mut signature = None;
+
+            let signature = D3D12SerializeRootSignature(
+                &desc,
+                D3D_ROOT_SIGNATURE_VERSION_1,
+                &mut signature,
+                &mut error,
+            )
+            .and_some(signature)?;
+
+            device.CreateRootSignature(0, signature.GetBufferPointer(), signature.GetBufferSize())
+        }
+    }
+
+    fn create_pipeline_state(
+        device: &ID3D12Device,
+        root_signature: &ID3D12RootSignature,
+    ) -> Result<ID3D12PipelineState> {
+        let compile_flags = if cfg!(debug_assertions) {
+            D3DCOMPILE_DEBUG | D3DCOMPILE_SKIP_OPTIMIZATION
+        } else {
+            0
+        };
+
+        let exe_path = std::env::current_exe().ok().unwrap();
+        let asset_path = exe_path.parent().unwrap();
+        let shaders_hlsl_path = asset_path.join("shaders.hlsl");
+        let shaders_hlsl = shaders_hlsl_path.to_str().unwrap();
+
+        let mut vertex_shader = None;
+        let vertex_shader = unsafe {
+            D3DCompileFromFile(
+                shaders_hlsl,
+                std::ptr::null_mut(),
+                None,
+                "VSMain",
+                "vs_5_0",
+                compile_flags,
+                0,
+                &mut vertex_shader,
+                std::ptr::null_mut(),
+            )
+            .and_some(vertex_shader)?
+        };
```

Comment:
Nit: I personally always try to keep as much safe code outside unsafe blocks as possible. Since `vertex_shader` is defined (and then shadowed) outside the `unsafe` block:


```suggestion
        }
        .and_some(vertex_shader)?;
```

Reply:
Multiline works well in the browser.

Reply:
Thanks - I am keen to iterate on style before this PR gets completed, so please do keep the suggestions coming!

Reply:
> Multiline works well in the browser.

Yup, using it all the time but wanted to try out the plugin for once. A PR to add that feature is on its way apparently :crossed_fingers:

Reply:
(oh, that's almost exactly the same example you posted below :) )

Reply:
That's a tricky one with the indentation change like that. It looks unusual but I'd probably stick with less-code-in-unsafe-is-better for consistency. Your pick though :)